File: Sources/MozzApp/Continuity/ContinuityCoordinator.swift
import Foundation
import MozzContinuity
import MozzCore
import MozzPlayback

/// A cross-device resume offer, ready to show the user.
public struct ContinuityOffer: Sendable, Identifiable {
    public var id: String { snapshot.cursor.current.remoteID + "@\(snapshot.cursor.capturedAtMS)" }
    public var snapshot: ContinuitySnapshot
    public var title: String
    public var artist: String
    /// Cover art for the track being offered, resolved against the active
    /// backend on this device — the stored queue carries only a reference, so
    /// the URL is minted locally and stays valid across token rotation.
    public var artwork: ArtworkRef?
    /// Device label, or nil where the backend cannot attribute a checkpoint —
    /// Subsonic's only signal is the client *product* name, which is identical
    /// for every Mozz install.
    public var deviceName: String?
    /// True when only the track and position survived: the queue could not be
    /// paired with the cursor, or the backend cannot store one.
    public var isTrackOnly: Bool
}

/// Drives cross-device playback continuity (ADR-0010).
///
/// Owns the two halves that must never be conflated:
/// - **Publishing** this device's checkpoint, debounced, so another device can
///   pick it up.
/// - **Offering** a remote checkpoint as "Continue here".
///
/// The governing rule is enforced here: nothing read from a store may stop
/// playback or make a playing device yield. A remote checkpoint only ever
/// becomes an *offer* the user can decline.
@MainActor
public final class ContinuityCoordinator: ObservableObject {
    @Published public private(set) var offer: ContinuityOffer?

    private var store: (any ContinuityStore)?
    private var deviceID: String = ""
    private var deviceName: String = ""

    /// Writes are suppressed until the first reconcile completes.
    ///
    /// Without this, cold launch restores the *local* session and would publish
    /// it immediately — overwriting a newer checkpoint another device wrote
    /// while this one was asleep. That is exactly the blind flush the ADR
    /// forbids.
    private var hasReconciled = false
    private var pending: (cursor: ContinuityCursor, queue: ContinuityQueue?)?
    private var flushTask: Task<Void, Never>?
    private var lastWrittenQueueHash: String?

    /// Checkpoints older than this are not offered. Subsonic can store neither a
    /// playback state nor a presence lease, so age is the only signal there.
    private let maxOfferAge: TimeInterval = 60 * 60 * 24 * 14

    public init() {}

    // MARK: Lifecycle

    /// Point the coordinator at a signed-in server, or clear it on sign-out.
    public func activate(store: (any ContinuityStore)?, deviceID: String, deviceName: String) {
        self.store = store
        self.deviceID = deviceID
        self.deviceName = deviceName
        hasReconciled = false
        offer = nil
        lastWrittenQueueHash = nil
    }

    /// Fetch the remote checkpoint and decide whether to offer it.
    ///
    /// Runs on activation and whenever the app returns to the foreground: an
    /// idle device has to refresh its view, otherwise a phone paused in a pocket
    /// while another device took over comes back showing stale state and
    /// clobbers the newer session on its next write.
    public func reconcile(isPlayingLocally: Bool) async {
        guard let store else {
            hasReconciled = true
            return
        }
        let loaded = try? await store.load()
        hasReconciled = true
        guard let snapshot = loaded else { return }
        let cursor = snapshot.cursor

        // Never offer to continue what this device itself last wrote.
        if !cursor.deviceID.isEmpty, cursor.deviceID == deviceID { return }
        if cursor.capturedAtMS > 0,
           Date().timeIntervalSince(cursor.capturedAt) > maxOfferAge { return }
        // A device that is actively playing is never interrupted — an offer
        // there would be inviting the user to stop what they are listening to.
        if isPlayingLocally { return }

        let item = snapshot.queue?.items.first {
            $0.locator.remoteID == cursor.current.remoteID
        }
        offer = ContinuityOffer(
            snapshot: snapshot,
            title: item?.title ?? "",
            artist: item?.artist ?? "",
            artwork: item?.artwork,
            deviceName: store.features.deviceAttribution && !cursor.deviceName.isEmpty
                ? cursor.deviceName
                : nil,
            isTrackOnly: snapshot.queue == nil
        )
    }

    public func dismissOffer() { offer = nil }

    // MARK: Publishing

    /// Record a checkpoint from the engine. Debounced; safe to call often.
    public func record(_ checkpoint: PlaybackCheckpoint, fingerprint: ServerAccountFingerprint) {
        guard let store, hasReconciled else { return }
        guard let current = checkpoint.queue.current else { return }

        let items = ContinuityMapper.items(from: checkpoint.queue, fingerprint: fingerprint)
        guard !items.isEmpty else { return }

        let queue = ContinuityQueueBuilder.make(
            items: items,
            descriptor: .adHoc,
            repeatMode: ContinuityMapper.repeatMode(checkpoint.queue.repeatMode),
            isShuffled: checkpoint.queue.isShuffled,
            totalCount: items.count
        )
        let cursor = ContinuityCursor(
            playbackRunID: checkpoint.runID,
            deviceID: deviceID,
            deviceName: deviceName,
            cursorSequence: checkpoint.sequence,
            capturedAtMS: Int64(Date().timeIntervalSince1970 * 1000),
            state: ContinuityMapper.state(checkpoint.status),
            current: TrackLocator(server: fingerprint, remoteID: current.id),
            currentAbsoluteIndex: max(checkpoint.queue.position, 0),
            positionMS: Int64(max(checkpoint.elapsed, 0) * 1000),
            queueHash: store.features.storesQueue ? queue.queueHash : nil
        )

        // Only resend the queue when it actually changed. Subsonic sends both in
        // one call regardless, but on Jellyfin this is what stops a 20-second
        // cursor write from dragging the whole queue across the network.
        let queueChanged = queue.queueHash != lastWrittenQueueHash
        pending = (cursor, queueChanged || store.features.truncatesQueue ? queue : nil)
        scheduleFlush(immediate: checkpoint.reason != .periodic)
    }

    private func scheduleFlush(immediate: Bool) {
        flushTask?.cancel()
        flushTask = Task { [weak self] in
            if !immediate {
                try? await Task.sleep(for: .seconds(3))
                if Task.isCancelled { return }
            }
            await self?.flush()
        }
    }

    private func flush() async {
        guard let store, let pending else { return }
        self.pending = nil
        do {
            try await store.save(pending.cursor, queue: pending.queue)
            if let queue = pending.queue { lastWrittenQueueHash = queue.queueHash }
        } catch {
            // Not worth surfacing or retrying: the next checkpoint carries
            // strictly newer state.
        }
    }

    /// Force any pending write out — used when backgrounding, which may be the
    /// last chance to run.
    public func flushNow() async {
        flushTask?.cancel()
        await flush()
    }
}

File: clients/android/app/src/main/java/com/thatcube/mozz/continuity/ContinuityCoordinator.kt
package com.thatcube.mozz.continuity

import android.util.Log
import com.thatcube.mozz.core.ContinuityDescriptor
import com.thatcube.mozz.core.ContinuityItemInput
import com.thatcube.mozz.core.ContinuityQueueInput
import com.thatcube.mozz.core.ContinuitySnapshot
import com.thatcube.mozz.core.MozzContinuity
import com.thatcube.mozz.core.ServerAccount
import com.thatcube.mozz.core.Track
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import java.util.UUID

/** Why a checkpoint is being written. */
enum class CheckpointReason { TRACK_CHANGED, SEEKED, TRANSPORT_CHANGED, QUEUE_CHANGED, PERIODIC }

/** A resume this device is willing to offer, already phrased for a person. */
data class ContinuityOffer(
    val snapshot: ContinuitySnapshot,
    val headline: String,
    val subtitle: String,
    val artworkKey: String?,
)

/**
 * Cross-device resume, from this phone's side (ADR-0010).
 *
 * Two halves that barely touch: publishing where this device got to, and
 * deciding whether what another device published is worth offering. Both are
 * deliberately incapable of stopping playback — a checkpoint carries resume
 * information and never ownership, which is what makes one shared slot with
 * last-writer-wins safe.
 *
 * Jellyfin and Subsonic hold the slot. Plex has no store for it: the core
 * answers that as a failure, and every call here treats a failure as "nothing
 * there" rather than as something to report. A Plex user should see no sign
 * this exists.
 */
class ContinuityCoordinator(
    private val continuity: MozzContinuity,
    private val deviceId: String,
    private val deviceName: String,
) {

    /** One write at a time: transitions fire faster than a round trip returns. */
    private val gate = Mutex()

    /** Identifies this run of playback. Re-minted whenever a new one starts. */
    private var runId = UUID.randomUUID().toString()
    private var sequence = 0L
    private var lastPeriodicMS = 0L

    /**
     * The hash of the queue last written, so an unchanged queue is not sent
     * again. A queue is the expensive half of a checkpoint and changes far
     * less often than a position does.
     */
    private var writtenQueueHash: String? = null

    /**
     * Nothing is published before the remote state has been read once.
     *
     * A phone paused in a pocket while another device took over would otherwise
     * come back and clobber the newer session with what it remembered.
     */
    private var reconciled = false

    fun beginRun() {
        runId = UUID.randomUUID().toString()
        sequence = 0
        lastPeriodicMS = 0
        writtenQueueHash = null
    }

    /**
     * Read the shared slot and decide whether to offer what is there.
     *
     * Returns null for every ordinary reason as well as every failure: no
     * checkpoint, this device's own, too old, or something is already playing
     * here.
     */
    suspend fun reconcile(
        serverId: String,
        isPlayingLocally: Boolean,
        nowMS: Long = System.currentTimeMillis(),
    ): ContinuityOffer? {
        val snapshot = runCatching { continuity.load(serverId) }.getOrNull()
        reconciled = true
        return offerFor(snapshot, deviceId, isPlayingLocally, nowMS)
    }

    /**
     * Write where this device has got to.
     *
     * Best effort throughout. A checkpoint that does not land is superseded by
     * the next one, and there is nothing a listener could do about a failure.
     */
    suspend fun checkpoint(
        reason: CheckpointReason,
        account: ServerAccount,
        queue: List<Track>,
        indexInQueue: Int,
        positionMS: Long,
        isPlaying: Boolean,
        repeatMode: String,
        isShuffled: Boolean,
        nowMS: Long = System.currentTimeMillis(),
    ) {
        if (!reconciled) return
        val current = queue.getOrNull(indexInQueue) ?: return
        // A steadily playing track re-checkpoints on a timer rather than on
        // every position tick, which arrives twice a second.
        if (reason == CheckpointReason.PERIODIC && nowMS - lastPeriodicMS < PERIODIC_INTERVAL_MS) return

        gate.withLock {
            if (reason == CheckpointReason.PERIODIC) lastPeriodicMS = nowMS
            sequence += 1

            val input = queueInput(account, queue, indexInQueue, repeatMode, isShuffled)
            val hash = runCatching { continuity.queueHash(input) }.getOrNull()?.queueHash
            // Send the queue only when it is not the one already stored. The
            // position moves constantly; the queue rarely does.
            val queueChanged = hash == null || hash != writtenQueueHash

            val saved = runCatching {
                continuity.save(
                    serverId = account.serverId,
                    playbackRunId = runId,
                    deviceId = deviceId,
                    deviceName = deviceName,
                    cursorSequence = sequence,
                    capturedAtMS = nowMS,
                    state = if (isPlaying) "playing" else "paused",
                    currentRemoteId = current.remoteId,
                    currentAbsoluteIndex = indexInQueue,
                    positionMS = positionMS,
                    queue = if (queueChanged) input else null,
                )
            }.onFailure {
                // Plex has no store, so this is the ordinary state there and
                // not worth a line every twenty seconds.
                Log.d(TAG, "no continuity store for ${account.serverName}")
            }.getOrNull()

            if (queueChanged) writtenQueueHash = saved?.queueHash ?: hash
        }
    }

    private fun queueInput(
        account: ServerAccount,
        queue: List<Track>,
        indexInQueue: Int,
        repeatMode: String,
        isShuffled: Boolean,
    ) = ContinuityQueueInput(
        descriptor = ContinuityDescriptor(kind = "queue"),
        items = queue.mapIndexed { index, track ->
            ContinuityItemInput(
                remoteId = track.remoteId,
                backend = account.kind.wire,
                // The server's own identity, not the row id derived from the
                // address it answers at today: a checkpoint has to survive the
                // same library being reached over a different route.
                serverId = account.machineIdentifier.orEmpty(),
                accountId = account.userId ?: account.username.orEmpty(),
                baseOrdinal = index,
                title = track.title,
                artist = track.artistName,
                durationMS = (track.durationSeconds * 1000).toLong(),
                artworkKey = track.artworkKey,
            )
        },
        repeatMode = repeatMode,
        isShuffled = isShuffled,
        totalCount = queue.size,
        startAbsoluteIndex = indexInQueue,
    )

    companion object {
        private const val TAG = "MozzContinuity"

        /** How often a steadily playing track re-checkpoints. Matches iOS. */
        const val PERIODIC_INTERVAL_MS = 20_000L

        /** Older than this and nobody is coming back to it. Matches both shells. */
        const val MAX_OFFER_AGE_MS = 14L * 24 * 60 * 60 * 1000

        /**
         * Whether a snapshot is worth putting in front of someone.
         *
         * The same four rules the iPhone and the desktop apply, in the same
         * order — this is a third statement of them, which is a drift risk
         * worth naming: see `ContinuityCoordinator.reconcile` in Swift and
         * `ContinuityPresentation.OfferFor` in C#.
         */
        fun offerFor(
            snapshot: ContinuitySnapshot?,
            localDeviceId: String,
            isPlayingLocally: Boolean,
            nowMS: Long,
        ): ContinuityOffer? {
            if (snapshot == null) return null
            val cursor = snapshot.cursor
            if (cursor.current.remoteId.isEmpty()) return null
            // Never offer to continue what this device itself last wrote.
            if (cursor.deviceId.isNotEmpty() && cursor.deviceId == localDeviceId) return null
            // A device already playing is never interrupted: an offer there is
            // an invitation to stop what you are listening to.
            if (isPlayingLocally) return null
            if (cursor.capturedAtMS > 0 && nowMS - cursor.capturedAtMS > MAX_OFFER_AGE_MS) return null

            val item = snapshot.queue?.items?.firstOrNull { it.locator.remoteId == cursor.current.remoteId }
            val hydrated = snapshot.hydratedTracks.firstOrNull { it.remoteId == cursor.current.remoteId }
            val title = item?.title?.ifBlank { null } ?: hydrated?.title
            val artist = item?.artist?.ifBlank { null } ?: hydrated?.artistName
            val subtitle = buildString {
                append(
                    when {
                        title == null -> "Track and position"
                        artist.isNullOrBlank() -> title
                        else -> "$title · $artist"
                    }
                )
                if (snapshot.queue == null || snapshot.isQueueMissing) append(" — track only")
            }
            val device = cursor.deviceName.trim().ifEmpty { "another device" }
            return ContinuityOffer(
                snapshot = snapshot,
                headline = "Resume from $device, ${age(cursor.capturedAtMS, nowMS)} ago",
                subtitle = subtitle,
                artworkKey = hydrated?.artworkKey ?: item?.artworkKey,
            )
        }

        /** "just now", "6 minutes", "2 hours", "3 days" — as on the desktop. */
        fun age(capturedAtMS: Long, nowMS: Long): String {
            if (capturedAtMS <= 0) return "just now"
            val seconds = ((nowMS - capturedAtMS) / 1000).coerceAtLeast(0)
            return when {
                seconds < 60 -> "just now"
                seconds < 3600 -> unit(Math.round(seconds / 60.0).toInt(), "minute")
                seconds < 86_400 -> unit(Math.round(seconds / 3600.0).toInt(), "hour")
                else -> unit(Math.round(seconds / 86_400.0).toInt(), "day")
            }
        }

        private fun unit(count: Int, noun: String) =
            if (count == 1) "1 $noun" else "$count ${noun}s"
    }
}
